File: cpp/model/Product.cpp
#include "Product.h"

Product::Product(const std::string& name, const ProductUnit& unit) : name(name), unit(unit) {}

std::string Product::getName() const {
    return name;
}

ProductUnit Product::getUnit() const {
    return unit;
}

bool Product::operator==(const Product& rhs) const {
    return name == rhs.name &&
           unit == rhs.unit;
}

bool Product::operator!=(const Product& rhs) const {
    return !(rhs == *this);
}

bool Product::operator<(const Product& rhs) const {
    if (name < rhs.name)
        return true;
    if (rhs.name < name)
        return false;
    return unit < rhs.unit;
}

bool Product::operator>(const Product& rhs) const {
    return rhs < *this;
}

bool Product::operator<=(const Product& rhs) const {
    return !(rhs < *this);
}

bool Product::operator>=(const Product& rhs) const {
    return !(*this < rhs);
}

File: php/src/Model/Product.php
<?php

declare(strict_types=1);

namespace Supermarket\Model;

use Ds\Hashable;

class Product implements Hashable
{
    public function __construct(
        private string $name,
        private ProductUnit $unit
    ) {
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getUnit(): ProductUnit
    {
        return $this->unit;
    }

    /**
     * @param Product $obj
     */
    public function equals($obj): bool
    {
        return $obj instanceof self &&
            $this->getName() === $obj->getName() &&
            $this->getUnit() === $obj->getUnit();
    }

    public function hash(): string
    {
        return "{$this->getName()}__{$this->getUnit()}";
    }
}
